ql/src/java/org/apache/hadoop/hive/ql/parse/MergeSemanticAnalyzer.java

```diff
@@ -295,72 +268,27 @@ private String chooseJoinType(List<ASTNode> whenClauses) {
    * cardinality_violation() UDF throws an error whenever it's called killing the query
    * @return true if another Insert clause was added
    */
-  private boolean handleCardinalityViolation(StringBuilder rewrittenQueryStr, ASTNode target,
-      String onClauseAsString, Table targetTable, boolean onlyHaveWhenNotMatchedClause, ColumnAppender columnAppender)
-              throws SemanticException {
+  private boolean handleCardinalityViolation(boolean onlyHaveWhenNotMatchedClause) {
```

Comment:
for COW that would be "AND EXISTS (SELECT ...)"

Reply:
Implemented Rewriter instead and added MergeStatement to store the parsed parts of merge. COWMergeRewiter should implement its own cardinality check logic

Reply:
👍